src/rp2_common/pico_rand/rand.c

```diff
@@ -0,0 +1,241 @@
+   rather suggest to use a xoroshiro128+ (for moderately parallel
+   computations) or xorshift1024* (for massively parallel computations)
+   generator.
+
+   Note:  This can be called with any value (i.e. including 0)
+*/
+static __noinline uint64_t splitmix64(uint64_t x) {
+    uint64_t z = x + 0x9E3779B97F4A7C15ull;
+    z = (z ^ (z >> 30)) * 0xBF58476D1CE4E5B9ull;
+    z = (z ^ (z >> 27)) * 0x94D049BB133111EBull;
+    return z ^ (z >> 31);
+}
+
+/* From the original source:
+
+   This is xoroshiro128** 1.0, one of our all-purpose, rock-solid,
+   small-state generators. It is extremely (sub-ns) fast and it passes all
+   tests we are aware of, but its state space is large enough only for
+   mild parallelism.
+
+   For generating just floating-point numbers, xoroshiro128+ is even
+   faster (but it has a very mild bias, see notes in the comments).
+
+   The state must be seeded so that it is not everywhere zero. If you have
+   a 64-bit seed, we suggest to seed a splitmix64 generator and use its
+   output to fill s.
+*/
+static inline uint64_t rotl(const uint64_t x, int k) {
+    return (x << k) | (x >> (64 - k));
+}
+
+static __noinline uint64_t xoroshiro128ss(void) {
+    const uint64_t s0 = rand_storage.r[0];
+    uint64_t s1 = rand_storage.r[1];
+
+    // Because the state is *modified* outside of this function, there is a
+    // 1 in 2^128 chance that it could be all zeroes (which is not allowed).
+    while (s0 == 0 && s1 == 0) {
+        s1 = time_us_64();   // should not be 0, but loop anyway
+    }
+
+    const uint64_t result = rotl(s0 * 5, 7) * 9;
+
+    s1 ^= s0;
+    rand_storage.r[0] = rotl(s0, 24) ^ s1 ^ (s1 << 16); // a, b
+    rand_storage.r[1] = rotl(s1, 37); // c
+
+    return result;
+}
+
+// Note: This can take many tens of milliseconds to execute
+static uint64_t sdbm_hash64_sram(uint64_t hash) {
+    // save some time by hashing a word at a time
+    for (uint i = (PICO_RAND_RAM_HASH_START + 3) & ~3; i < PICO_RAND_RAM_HASH_END; i+=4) {
+        uint32_t c = *(uint32_t *) i;
+        hash = (uint64_t) c + (hash << 6) + (hash << 16) - hash;
+    }
+    return hash;
+}
+
+// check that the ROSC is running but that the processors are NOT running from it
+static inline void check_rosc_asserts(void) {
+#if !PICO_RAND_DISABLE_ROSC_CHECKS
+    hard_assert(rosc_hw->status & ROSC_STATUS_ENABLED_BITS);
+    hard_assert((clocks_hw->clk[clk_sys].ctrl & CLOCKS_CLK_SYS_CTRL_AUXSRC_BITS) !=
+                (CLOCKS_CLK_SYS_CTRL_AUXSRC_VALUE_ROSC_CLKSRC << CLOCKS_CLK_SYS_CTRL_AUXSRC_LSB));
+#endif
+}
+
+static void update_rosc_samples_under_lock(void) {
+    static uint32_t previous_sample_time_us = 0;
+
+    check_rosc_asserts();
+
+    // Ensure that the ROSC random bit is not sampled too quickly,
+    // ROSC may be ticking only a few times a microsecond.
+    // Note: In general (i.e. sporadic) use, very often there will be no delay here.
+    while ((time_us_32() - previous_sample_time_us) < PICO_RAND_MIN_ROSC_SAMPLE_TIME_US) {
```

Comment:
If the ROSC **isn't** enabled, is there any point `in having this delay? (Or even in calling `update_rosc_samples_under_lock` at all?)  But perhaps tweaking that is an optimisation that's not worth it.